### ISSUE TITLE
Typeclass derivation of recursive types

### ISSUE THREAD

**Issue by mpilquist** [scala/scala3#8183]
The example `Eq` typeclass derivation fails on recursive types -- either in trying to reacquire a lazy val for directly recursive types (`Rec` below) or with a `StackOverflow` on mutual recursive types (`Tree` below).

## minimized code

```Scala
import scala.deriving._
import scala.compiletime._

inline def summon[T]: T = summonFrom { case t: T => t }

inline def summonAll[T <: Tuple]: List[Eq[_]] = inline erasedValue[T] match {
  case _: Unit => Nil
  case _: (t *: ts) => summon[Eq[t]] :: summonAll[ts]
}

trait Eq[T] {
  def eqv(x: T, y: T): Boolean
}

object Eq {
  given Eq[Int] {
    def eqv(x: Int, y: Int) = x == y
  }

  def check(elem: Eq[_])(x: Any, y: Any): Boolean =
    elem.asInstanceOf[Eq[Any]].eqv(x, y)

  def iterator[T](p: T) = p.asInstanceOf[Product].productIterator

  def eqSum[T](s: Mirror.SumOf[T], elems: List[Eq[_]]): Eq[T] =
    new Eq[T] {
      def eqv(x: T, y: T): Boolean = {
        val ordx = s.ordinal(x)
        (s.ordinal(y) == ordx) && check(elems(ordx))(x, y)
      }
    }

  def eqProduct[T](p: Mirror.ProductOf[T], elems: List[Eq[_]]): Eq[T] =
    new Eq[T] {
      def eqv(x: T, y: T): Boolean =
        iterator(x).zip(iterator(y)).zip(elems.iterator).forall {
          case ((x, y), elem) => check(elem)(x, y)
        }
    }

  inline given derived[T](given m: Mirror.Of[T]): Eq[T] = {
    val elemInstances = summonAll[m.MirroredElemTypes]
    inline m match {
      case s: Mirror.SumOf[T]     => eqSum(s, elemInstances)
      case p: Mirror.ProductOf[T] => eqProduct(p, elemInstances)
    }
  }
}

object Main extends App {
  
  case class Rec(x: Int, y: Option[Rec]) derives Eq
  println(summon[Eq[Rec]])
  /* Deriving Eq[Rec] hangs with stack like:
     java.lang.Thread.State: WAITING (on object monitor)
        at java.lang.Object.wait(java.base@11.0.1/Native Method)
        - waiting on <0x00000007fd82c240> (a java.lang.Object)
        at java.lang.Object.wait(java.base@11.0.1/Object.java:328)
        at dotty.runtime.LazyVals$.wait4Notification(LazyVals.scala:89)
        - waiting to re-lock in wait() <0x00000007fd82c240> (a java.lang.Object)
        at rs$line$1$Main$Rec$.derived$Eq(rs$line$1:59)
        at rs$line$1$Main$Rec$.derived$Eq(rs$line$1:8)
    */
  
  enum Tree[A] derives Eq {
    case Node[A](l: Tree[A], r: Tree[A]) extends Tree[A]
    case Leaf[A](value: A) extends Tree[A]
  }
  // Deriving Eq[Tree[Int]] throws SOE
  println(summon[Eq[Tree[Int]]])
}
```

Aside: the example defines `derived` as a `given` instead of a `def` but doing so seems to make the `derives` clause redundant / unnecessary.

**Comment by mpilquist** [comment]
Note: the implementation of Eq in [typeclass-derivation3.scala](https://github.com/lampepfl/dotty/blob/98103a216ca00658631e7c0abd35f2929dc5e80a/tests/run/typeclass-derivation3.scala#L29-L80) works fine for recursive types. Maybe close this as won't fix and update the docs to use that implementation, along with a warning about directly summoning recursive instances?

**Comment by biboudis** [comment]
I will keep it a bit more open as I am currently investigating this. I think your original post reports a legit bug but thanks for reporting back Michael.

**Comment by kubukoz** [comment]
I have an example for a big simpler type, hanging in runtime:

```scala
import scala.deriving.Mirror
import scala.compiletime.erasedValue
import scala.compiletime.summonInline
import scala.compiletime.summonFrom

object Derivations {
  inline def summonAll[T <: Tuple]: List[Print[_]] = inline erasedValue[T] match {
    case _: Unit => Nil
    case _: (t *: ts) => summonInline[Print[t]] :: summonAll[ts]
  }

  inline def gen[T](using m: Mirror.Of[T]): Print[T] = {
    val elems = summonAll[m.MirroredElemTypes]
    println("foo")

    _ => "bar"
  }
}

trait Print[T] {
  def (t: T).print: String
}

object Print {
  inline implicit def derived[T: Mirror.Of]: Print[T] = Derivations.gen[T]

  given Print[Int] = _.toString
}

enum MyList derives Print {
  case Cons(h: Int, t: MyList)
  case End
}

import Print.{given _}
@main
def run = println(MyList.Cons(1, MyList.Cons(2, MyList.End)).print)
```

(hangs between printing foo and printing bar)

**Comment by cb372** [comment]
Can anyone explain why the implementation @mpilquist linked *does* work for recursive types? It looks pretty similar to the example in the documentation, which doesn't work.

By the way, I worked around the problem by introducing a wrapper type so I could defer implicit resolution using `=>`. I'll leave it here just in case it helps anyone: https://github.com/cb372/type-class-derivation-in-scala-3/blob/1fe166ea13ce6ff09740311e47f72b655ca9b373/src/main/scala/mylibrary/Functor.scala#L67-L80

**Comment by dwijnand** [comment]
> Can anyone explain why the implementation @mpilquist linked _does_ work for recursive types? It looks pretty similar to the example in the documentation, which doesn't work.

I think it's the `summonFrom` block that adds a level of delay/deferred-ness (by being a compile-time conditional I think).  Replacing it with `summonInline` breaks things:

```scala
        if (ord == n) {
          val m = summonInline[Mirror.ProductOf[t]]
          eqlElems[m.MirroredElemTypes](0)(x.asInstanceOf[Product], y.asInstanceOf[Product])
          //summonFrom {
          //  case m: Mirror.ProductOf[`t`] => eqlElems[m.MirroredElemTypes](0)(x.asInstanceOf[Product], y.asInstanceOf[Product])
          //}
        } else eqlCases[ts](n + 1)(x, y, ord)
```

```scala
[error] -- Error: /d/play-json/play-json/shared/src/main/scala-3/play/api/libs/json/JsMacroImpl.scala:14:23
[error] 14 |  enum Lst[+T] derives Eq {
[error]    |                       ^
[error]    |cannot reduce inline match with
[error]    | scrutinee:  scala.compiletime.erasedValue[m.MirroredElemTypes] : m.MirroredElemTypes
[error]    | patterns :  case _:*:[t @ _, ts @ _]
[error]    |             case _:EmptyTuple
[error]    | This location contains code that was inlined from JsMacroImpl.scala:30
[error]    | This location contains code that was inlined from JsMacroImpl.scala:40
[error]    | This location contains code that was inlined from JsMacroImpl.scala:49
```

**Comment by kubukoz** [comment]
3.0.0 still does this.

Updated code after syntax changes:

```scala
import scala.deriving.*
import scala.compiletime.*

inline def summonAll[T <: Tuple]: List[Eq[_]] = inline erasedValue[T] match {
  case _: EmptyTuple => Nil
  case _: (t *: ts)  => summonInline[Eq[t]] :: summonAll[ts]
}

trait Eq[T] {
  def eqv(x: T, y: T): Boolean
}

object Eq {
  given Eq[Int] with
    def eqv(x: Int, y: Int) = x == y

  def check(elem: Eq[_])(x: Any, y: Any): Boolean =
    elem.asInstanceOf[Eq[Any]].eqv(x, y)

  def iterator[T](p: T) = p.asInstanceOf[Product].productIterator

  def eqSum[T](s: Mirror.SumOf[T], elems: List[Eq[_]]): Eq[T] =
    new Eq[T] {

      def eqv(x: T, y: T): Boolean = {
        val ordx = s.ordinal(x)
        (s.ordinal(y) == ordx) && check(elems(ordx))(x, y)
      }

    }

  def eqProduct[T](p: Mirror.ProductOf[T], elems: List[Eq[_]]): Eq[T] =
    new Eq[T] {

      def eqv(x: T, y: T): Boolean =
        iterator(x).zip(iterator(y)).zip(elems.iterator).forall { case ((x, y), elem) =>
          check(elem)(x, y)
        }

    }

  inline given derived[T](
    using m: Mirror.Of[T]
  ): Eq[T] = {
    val elemInstances = summonAll[m.MirroredElemTypes]
    inline m match {
      case s: Mirror.SumOf[T]     => eqSum(s, elemInstances)
      case p: Mirror.ProductOf[T] => eqProduct(p, elemInstances)
    }
  }

}

object Main extends App {

  case class Rec(x: Int, y: Option[Rec]) derives Eq
  println(summon[Eq[Rec]])

  /* Deriving Eq[Rec] hangs with stack like:
     java.lang.Thread.State: WAITING (on object monitor)
        at java.lang.Object.wait(java.base@11.0.1/Native Method)
        - waiting on <0x00000007fd82c240> (a java.lang.Object)
        at java.lang.Object.wait(java.base@11.0.1/Object.java:328)
        at dotty.runtime.LazyVals$.wait4Notification(LazyVals.scala:89)
        - waiting to re-lock in wait() <0x00000007fd82c240> (a java.lang.Object)
        at rs$line$1$Main$Rec$.derived$Eq(rs$line$1:59)
        at rs$line$1$Main$Rec$.derived$Eq(rs$line$1:8)
   */

  enum Tree[A] derives Eq {
    case Node[A](l: Tree[A], r: Tree[A]) extends Tree[A]
    case Leaf[A](value: A) extends Tree[A]
  }

  // Deriving Eq[Tree[Int]] throws SOE
  println(summon[Eq[Tree[Int]]])
}
```

```
Name: main
State: WAITING on java.lang.Object@12840f8c
Total blocked: 0  Total waited: 1

Stack trace: 
java.base@11.0.10/java.lang.Object.wait(Native Method)
java.base@11.0.10/java.lang.Object.wait(Object.java:328)
app//scala.runtime.LazyVals$.wait4Notification(LazyVals.scala:89)
app//Main$Rec$.derived$Eq(Examples.scala:102)
app//Main$Rec$.t$1(Examples.scala:102)
app//Main$Rec$.derived$Eq(Examples.scala:102)
app//Main$.<clinit>(Examples.scala:103)
app//Main.main(Examples.scala)
```

**Comment by pmeheut** [comment]
Any chance of seing this fixed? It has been more than 2 years.
For instance, my caliban based code used to compile perfectly but after simply upgrading the libraries to more recent minor versions, I get something very cryptic and impossible to fix:

```
error]     |           method derived is declared as `inline`, but was not inlined
[error]     |
[error]     |           Try increasing `-Xmax-inlines` above 32
[error]     |---------------------------------------------------------------------------
[error]     |Inline stack trace
[error]     |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
[error]     |This location contains code that was inlined from Derivation.scala:16
[error]     |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
[error]     |This location contains code that was inlined from Derivation.scala:16
[error]     |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
[error]     |This location contains code that was inlined from Derivation.scala:16
[error]     |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
[error]     |This location contains code that was inlined from Derivation.scala:16
[error]     |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
[error]     |This location contains code that was inlined from Derivation.scala:16
[error]     |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
[error]     |This location contains code that was inlined from Derivation.scala:16
```

**Comment by bishabosha** [comment]
I recently updated the [documentation](https://docs.scala-lang.org/scala3/reference/contextual/derivation.html) (2 months ago) to use a new algorithm that handles recursive types - I wasn't aware of this issue, could you try it?

**Comment by pmeheut** [comment]
Thanks. After reading it, I'm not sure this is relevant to my problem.
I'll open another issue.

**Comment by MartinHH** [comment]
> I recently updated the [documentation](https://docs.scala-lang.org/scala3/reference/contextual/derivation.html) (2 months ago) to use a new algorithm that handles recursive types - could you try it then we can hopefully close this?

I followed that example to improve some derivation code's support for recursion and noticed that it only works if the recursion happens  within the sum which is recurred to. It fails if the recursion is nested somewhere deeper in one of the members of the sum. For example, this fails:

```
case class Tail[+T](ts: Lst[T])

// compilation fails with:
// [error] No given instance of type Eq[Tail[T]] was found
// [error] enum Lst[+T] derives Eq:
enum Lst[+T] derives Eq:
  case Cns(t: T, tail: Tail[T])
  case Nl
```

I managed to come up with a [solution](https://github.com/MartinHH/recursive_derivation_example/commit/1e005753c3c1ba0fc32e9d09fb858ab81e5c0c8c#diff-939662c2a63a839cd04c9bce3541f36a01d04239446f7f7fc99a55444a64873dR27).

However: I would not be able to explain why this solution works while another, simpler approach doesn't (see my [comment](https://github.com/MartinHH/recursive_derivation_example/commit/1e005753c3c1ba0fc32e9d09fb858ab81e5c0c8c#r124926420) here).

**Comment by jchyb** [comment]
I tested @bishabosha's updated docs example and it works well. For:
```scala
 case class Rec(x: Int, y: Option[Rec]) derives Eq
```
it correctly responds with an error about lack of `Eq` for `Option[Rec]`, and for
```scala
enum Tree[A] derives Eq {
  case Node[A](l: Tree[A], r: Tree[A]) extends Tree[A]
  case Leaf[A](value: A) extends Tree[A]
}
```
It successfully derives the typeclass (both examples taken from the original minimisation). 
With all that I feel comfortable closing this issue. Let me know if I missed something in the discussion above.
